test: add semgrep-core-proprietary binary gha artifact cache

amchiclet opened the pull request:
For some kinds of changes, we need the pro binary, so we can make sure that pytests work correctly in pro mode.

Example PR that would benefit from this change: [enabling interproc taint scans by default](https://github.com/semgrep/semgrep/pull/9872) or [adding pro flags to the metrics test](https://github.com/semgrep/semgrep/pull/9258).

PRs that would not benefit from this change are PRs in semgrep that require changes from Pro. The workflow brings in the develop branch, so WIP in pro would not be brought in.

Eventually, we are headed towards a monorepo. If that is coming soon, then this PR can probably be ignored.

This PR extends the `check-semgrep-pro` workflow to also upload the binaries (both semgrep-core and semgrep-core-proprietary) to gh cache. The semgrep-core-proprietary is built using code from develop and includes the most recent changes here.

I needed to change the container to alpine to force a statically linked binary. I haven't found a way to statically link on ubuntu yet. We need a statically linked binary so that the tests don't need to depend on the tree sitter library being installed. (This is also the case with semgrep-core that is built in alpine too.)

When downloading the binary, also needed to make sure `pro-installed-by.txt` has the correct info and is in the same directory as the binary. Otherwise, the CLI won't use the binary.

Added a test `test_taint_intrafile.py` which uses the `--pro_intrafile` flag and the test passes. No other tests requiresthe pro binary.

amchiclet committed:
chore(ci): upload semgrep-core-proprietary to gh artifacts cache

amchiclet commented:
I actually just learned about this file: https://github.com/semgrep/semgrep/blob/develop/src/main/flags.sh. It's probably possible to specify an opam switch, etc and make the build static on ubuntu too.

The good thing about alpine is that it's consistent with how semgrep-core is built in build-test-core-x86, but maybe it's build-test-core-x86 that should move to ubuntu? Not sure.

emjin commented on .github/workflows/tests.jsonnet:
```diff
@@ -220,6 +221,40 @@ local install_x86_artifacts = {
+    name: core_pro_x86.export.artifact_name,
+  },
+};
+local install_x86_pro_artifacts = {
+  name: 'Install pro artifacts',
+  run: |||
+    tar xf ocaml-build-artifacts.tgz
+    sudo cp ocaml-build-artifacts/bin/* /usr/bin
+
+    # The version is stored in cli/src/semgrep/__init__.py
+    # and the file content looks like
+    #   __VERSION__ = "x.xx.x"
+    # so we try to get the x.xx.x part from that file below.
+
+    # First, read the file content
+    version=$(<cli/src/semgrep/__init__.py)
+
+    # Next, we remove the prefix that leads up to the number.
+    version="${version##__VERSION__ = \"}"
+
+    # Finally, we remove the final quote suffix.
+    version="${version%*\"}"
+
+    # By writing the version number to this file and putting it
+    # in the same directory as semgrep-core-proprietary,
+    # the binary will pass the cli's checks and can be run.
+    echo "$version" > pro-installed-by.txt
+    sudo cp pro-installed-by.txt /usr/bin
+  |||,
+};
```
Intuitively it seems like it would be easier to use `--custom-binary` like we do in the `semgrep-proprietary` tests? If this is actually hard for some reason I think a comment would be nice to share the knowledge

amchiclet replied:
Oooh, I did not know about `--custom-binary`. I'll try and report back!

emjin commented on .github/workflows/tests.yml:
```diff
@@ -197,7 +200,7 @@ jobs:
     name: test semgrep-cli
     needs:
-      - build-test-core-x86
+      - check-semgrep-pro
```
I am curious how much time this adds to our test process, if you happen to know

amchiclet replied:
The short answer is that the whole `tests` workflow takes the same amount of time compared to before.

The tests themselves take roughly the same time because it's just downloading a different artifact, and the scan still hits the OSS version. Once we also run pytest for e2e-pro in CI, maybe that will change.

The build takes longer (before 9m, now 11m) because it's built in a different environment (alpine vs ubuntu).

Regardless, the `build-test-windows-x86` and the `test-semgrep-pro` jobs both [dominate the build time for this whole workflow](https://github.com/semgrep/semgrep/actions/runs/8173081223), so this pro build on x86 is masked, and the overall process takes the same amount of time.

At some point I feel like many workflows are doing similar things, and we may be able to consolidate. I just looked at `test-semgrep-pro` and `check-semgrep-pro` are similar but they're not exactly doing the same things. But maybe there are some similarities after all, and we can save some time.

emjin commented on .github/workflows/check-semgrep-pro.jsonnet:
```diff
@@ -7,53 +7,60 @@ local gha = import 'libs/gha.libsonnet';
-  // which is needed to checkout semgrep-pro from semgrep GHA.
-  // alt: use our r2c alpine-ocaml and procedure to download gh binary tarball
+local job = container.job {
+  // The ocaml_alpine container we are using here has opam already
+  // installed, as well as an opam switch already created, and a big set of
+  // packages already installed. Thus, the
+  // 'make install-deps-ALPINE-for-semgrep-core' below is very fast and almost a
+  // noop.
```
This explains why it's ok that we run `make install-deps-ALPINE-for-semgrep-core` but I missed why we added that step. (I assume it's related to static linking as mentioned in the PR---but I'm still a bit confused about why that's an issue. Also, it's helpful to have a comment in the workflow since as a team we have less knowledge of github actions)

amchiclet replied:
I see what you're saying and I'll add more comments.

Basically the docker image is a blackbox and I'm not sure if we can really expect all the required things for an alpine build to be there.

emjin reviewed (commented):
Left some initial questions. We should have a small test with this PR so that we exercise the pro binary

aryx reviewed (changes requested):
Supportive of this change!
Seems like a good idea. We just need to support both the cli tests with both semgrep-core oss and semgrep-core-proprietary.

amchiclet committed:
try custom-binary flag

IagoAbal commented on .github/workflows/check-semgrep-pro.jsonnet:
```diff
@@ -7,53 +7,60 @@ local gha = import 'libs/gha.libsonnet';
 local actions = import 'libs/actions.libsonnet';
 local semgrep = import 'libs/semgrep.libsonnet';
 
+// exported for other workflows
```
Why not just porting the tests that we want to run to the Pro repo? We need to do that in preparation for the monorepo. I haven't reviewed this carefully, but if you're just extending `check-semgrep-pro`, then I'm guessing that this won't work for PRs that require updates to Pro, which are many.

emjin replied:
I did that in the past, but I'm not a fan of duplicating the pytest test infrastructure. Since `--pro-intrafile` will be default, I want all the semgrep ci tests to be run using `--pro-intrafile`. If someone adds a `semgrep ci` test, I don't want them to have to go to a different repo to make it use the correct binary. @kopecs also had some tests he wanted to add that made more sense in the semgrep repo.

> This will not work when the PR requires another PR in Pro

That's true, but when that happens CI is red anyway usually

amchiclet replied:
Sounds like the bottom line is that there are definitely some tests that make sense to be in the semgrep repo.

But the contention is that we are moving to monorepo soon (IIUC?), and we are worried about doing work that we will eventually throw away or make the migration harder. For this point, I still this PR is still the right thing to do now. I feel like mono repo may take a while. I'm being pessimistic here though.

I don't understand the issue about a semgrep PR requiring a PR from Pro. I'm only familiar with the other way around. What's an example?

IagoAbal replied:
> That's true, but when that happens CI is red anyway usually

Yes, but then we merge the PR regardless of the CI being red.

How much do we need to duplicate? Nowadays we use Jsonnet to generate the CI workflows `.yml` files. Couldn't we use Jsonnet to reduce duplication to the minimum? Also, if we set this up, it won't be duplicated for too long, both @aryx and I really want to switch to the monorepo ASAP. And the CI workflows are the main blocker, if we move them, we could switch to a monorepo fairly soon I think.

emjin replied:
> How much do we need to duplicate?

I'm not quite sure, actually. Is your vision that the pytests are all entirely duplicated on the semgrep-proprietary side? And that we add new pytests for pro in the semgrep repo? I think that wouldn't be too much effort, if it's just a mirror, and I could be convinced to support that. My main concern is that sometimes you might not want to add a pytest on the semgrep side for various reasons (e.g. testing a new rule format SR wants to keep quiet)

IagoAbal replied:
If we want to do things well, I think there is no choice than to run for each Pro PR all the checks we run for OSS PRs. (This is actually regardless of having a monorepo or not having it.) If we disagree on this point we should have a video call Um, you and I, because I would like to understand your point of view, I could be wrong.

amchiclet replied:
> run for each Pro PR all the checks we run for OSS PRs

I agree with this direction.

Regarding this PR in particular, what do you think is a good way to test changes like [making pro-intrafile turned on by default](https://github.com/semgrep/semgrep/pull/9872)?

Should it be a manual trigger from the pro repo instead, where the dev can insert a PR to test?

IagoAbal commented:
> Supportive of this change! Seems like a good idea. We just need to support both the cli tests with both semgrep-core oss and semgrep-core-proprietary.

I think it's better to setup all these tests in the Pro repo, which we need to do anyways for the monorepo. This will not work when the PR requires another PR in Pro, which at least in my case is "99%" of the times.

amchiclet committed:
added a bit more comments in check-semgrep-pro

amchiclet committed:
try a different directory for the pro binary

amchiclet committed:
added pro test

amchiclet committed:
install to tmp-bin

amchiclet committed:
added snapshot

amchiclet committed:
fix path

amchiclet committed:
marked the intrafile test as kinda slow

amchiclet committed:
rename function

amchiclet committed:
add app token to simulate login

amchiclet commented:
Added a small test. Rebased. It's calling Pro, but got some snapshot error and also some 404 errors which didn't happen locally. Will continue investigating tomorrow.

IagoAbal commented on .github/workflows/tests.jsonnet:
```diff
@@ -220,6 +221,29 @@ local install_x86_artifacts = {
+    # we don't need to run semgrep with sudo.
+    mv ocaml-build-artifacts/bin cli/tmp-bin
+    cd cli
+    PATH="tmp-bin:$PATH" pipenv run semgrep install-semgrep-pro --custom-binary tmp-bin/semgrep-core-proprietary
```
Can you update the PR description and specify what kind of issues you want to find and what issues you will not find this way?

amchiclet replied:
Done.

IagoAbal commented on .github/workflows/tests.jsonnet:
```diff
@@ -578,8 +605,9 @@ local ignore_md = {
     'test-semgrep-core': test_semgrep_core_job,
     'test-osemgrep': test_osemgrep_job,
-    // Pysemgrep tests, requires build-test-core-x86 job
+    // Pysemgrep tests that require check-semgrep-pro
```
Is any test using the Pro binary except for the new one you added here `test_taint_intrafile` ?

amchiclet replied:
No other test here is using the Pro binary.

Just future tests in this PR: https://github.com/semgrep/semgrep/pull/9872

amchiclet committed:
add snapshot for intrafile

amchiclet committed:
remove wrong snapshot file

amchiclet committed:
move e2e-pro-need-login to its own directory

amchiclet committed:
update test to run only pro that needs login

amchiclet committed:
make linting pass

amchiclet committed:
remove the need for token

amchiclet committed:
trim white space

amchiclet committed:
fix precommit

amchiclet committed:
whitespaces included

amchiclet committed:
move e2e need login tests

amchiclet committed:
update pre-commit settings

amchiclet commented:
All tests are passing again.

Two PRs would benefit from this AFAIK:
* https://github.com/semgrep/semgrep/pull/9872
* https://github.com/semgrep/semgrep/pull/9258

If we decide not to move forward with this PR, I'm kinda ok with the time spent because is was a useful learning experience anyway.

I'll just be a bit sad it's blocking these 2 PRs.

amchiclet commented:
> We just need to support both the cli tests with both semgrep-core oss and semgrep-core-proprietary.

@aryx  I'm not sure what this means. Would you like more than just adding a pro test here?
